File: scripts/algorithms/H/H-Index II/H-Index II.cpp
// Runtime: 23 ms (Top 74.27%) | Memory: 18.7 MB (Top 49.37%)
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int start = 0 , end = citations.size()-1;
        int n = citations.size();
        while(start <= end){
            int mid = start + (end - start) / 2;
            int val = citations[mid];
            if(val == (n - mid)) return citations[mid];
            else if(val < n - mid){
                start = mid + 1;
            }
            else{
                end = mid - 1;
            }
        }
        return n - start;
    }
};

File: scripts/algorithms/H/HIndex II/HIndex II.js
// Runtime: 49 ms (Top 77.42%) | Memory: 43.00 MB (Top 96.77%)

var hIndex = function(citations) {
        
    let start = 0, end = citations.length-1;
    
    while (start <= end) {
        
        let mid = Math.floor((start+end)/2);
        
        if (citations.length - mid - 1 < citations[mid]) {
            
            end = mid - 1;
        } else {
            
            start = mid + 1;
        }
    }
    
    return citations.length - start;
};
